net: lib: call socket poll if could not send data completely in http or ws (websocket)

AndreyDodonov-EH opened the pull request:
Revived [PR](https://github.com/zephyrproject-rtos/zephyr/pull/45633) 
+ introduce changes suggested by @jukkar there (respect timeout supplied by the user).

In fact timeouts might not be completely respected if net_context timeout is big enough, but it is still better than nothing.

In http_client and websocket, if we couldn't send all (or any data) via the socket,
invoke poll instead of blindly retrying and flooding the socket.

This fix is especially relevant after adding https://github.com/zephyrproject-rtos/zephyr/pull/45626

jukkar commented on subsys/net/lib/http/http_client.c:
```diff
@@ -29,12 +29,32 @@ LOG_MODULE_REGISTER(net_http_client, CONFIG_NET_HTTP_LOG_LEVEL);
-static int sendall(int sock, const void *buf, size_t len)
+static int sendall(int sock, const void *buf, size_t len, int32_t *timeout)
 {
+	int64_t timestamp = k_uptime_get();
```
There is the `sys_timepoint` API we could use here, it is used in the system for this purposes (to calculate the timeouts)

AndreyDodonov-EH replied:
@jukkar I just tried to stay consistent with what was there, but I guess you are right, I'll check out `sys_timepoint` API and will replace the calls accordingly.

AndreyDodonov-EH replied:
@jukkar You are right, it's indeed cleaner.
I just passed down a `k_timepoint_t` by which the function should add or timeout.

The only downside is that I had to convert then `k_timeout` to ms, since `zsock_poll` expects ms (which is correct, since posix `poll` expects milliseconds).

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -29,12 +29,28 @@ LOG_MODULE_REGISTER(net_http_client, CONFIG_NET_HTTP_LOG_LEVEL);
 
-static int sendall(int sock, const void *buf, size_t len)
+static int sendall(int sock, const void *buf, size_t len,
+					const k_timepoint_t req_end_timepoint)
```
nit: alignment

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -566,6 +580,7 @@ int http_client_req(int sock, struct http_request *req,
 	method = http_method_str(req->method);
 
 	ret = http_send_data(sock, send_buf, send_buf_max_len, &send_buf_pos,
+				 req_end_timepoint,
```
Alignment, could rearrange arguments to better fit into lines.

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -616,6 +631,7 @@ int http_client_req(int sock, struct http_request *req,
 		     i++) {
 			ret = http_send_data(sock, send_buf, send_buf_max_len,
 					     &send_buf_pos,
+						 req_end_timepoint,
```
Alignment, but I'd just move this to the line above.

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -119,13 +136,14 @@ static int http_send_data(int sock, char *send_buf,
 
-static int http_flush_data(int sock, const char *send_buf, size_t send_buf_len)
+static int http_flush_data(int sock, const char *send_buf, size_t send_buf_len,
+							const k_timepoint_t req_end_timepoint)
```
Alignment

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -29,12 +29,28 @@ LOG_MODULE_REGISTER(net_http_client, CONFIG_NET_HTTP_LOG_LEVEL);
+			pfd.fd = sock;
+			pfd.events = ZSOCK_POLLOUT;
+			pollres = zsock_poll(&pfd, 1, req_timeout_ms);
+			if (pollres >= 0) {
```
Shouldn't `pollres == 0` break the loop? Currently if we reach the timeout, it'd just start busy-looping.

AndreyDodonov-EH replied:
You are right. 
I'll correct it to 
```
pollres = zsock_poll(&pfd, 1, req_timeout_ms);
if (pollres == 0) {
    return -ETIMEDOUT;
}
else if (pollres > 0) {
    continue;
} else {
    return -errno;
}
```

rlubos commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -556,7 +557,23 @@ static int sendmsg_all(int sock, const struct msghdr *message, int flags)
+			pfd.fd = sock;
+			pfd.events = ZSOCK_POLLOUT;
+			pollres = zsock_poll(&pfd, 1, req_timeout_ms);
+			if (pollres >= 0) {
```
Ditto.

rlubos commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -544,7 +544,8 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
 #if !defined(CONFIG_NET_TEST)
-static int sendmsg_all(int sock, const struct msghdr *message, int flags)
+static int sendmsg_all(int sock, const struct msghdr *message, int flags,
+						const k_timepoint_t req_end_timepoint)
```
Alignment

AndreyDodonov-EH committed:
net: lib: http: call socket poll for http_client send

If we couldn't send all (or any data) via the socket,
invoke poll instead of blindly retrying and flooding the socket.
Respect timeout through http_client_req

Signed-off-by: Andrey Dodonov <Andrey.Dodonov@endress.com>

AndreyDodonov-EH committed:
net: lib: websocket: call socket poll for websocket

If we couldn't send all (or any data) via the socket,
invoke poll instead of blindly retrying and flooding the socket

Signed-off-by: Andrey Dodonov <Andrey.Dodonov@endress.com>

pdgendt commented on subsys/net/lib/http/http_client.c:
```diff
@@ -29,12 +29,30 @@ LOG_MODULE_REGISTER(net_http_client, CONFIG_NET_HTTP_LOG_LEVEL);
 #define MAX_SEND_BUF_LEN 192
 
-static int sendall(int sock, const void *buf, size_t len)
+static int sendall(int sock, const void *buf, size_t len,
+			const k_timepoint_t req_end_timepoint)
```
Alignment of some functions still off...

AndreyDodonov-EH replied:
@pdgendt 
I couldn't find any clear rules in Zephyr Contributing quidelines on indentation.
De facto, though, indentation should be done with tabs (otherwise CI throws an error).
So here just with tabs you can't always align next line of parameters properly.

So if you have a suggestion here (apart from using spaces), I'm open for a change

rlubos replied:
> De facto, though, indentation should be done with tabs (otherwise CI throws an error).
So here just with tabs you can't always align next line of parameters properly.

Finetuning indentation with spaces does not throw an error (I. e. use spaces instead of the last tab in the line). Sure it does not seem to be covered by the coding guidelines, but we tend to align to the opening brace all over the codebase, it simply looks neater.

pdgendt replied:
So you have to indent with tabs, and align with the opening brace using spaces.

AndreyDodonov-EH replied:
Okay, will do!

AndreyDodonov-EH replied:
Too late, it got merged... But I'll keep in mind for the next PR, thanks!
